test/test_filesystem.cpp

```diff
@@ -541,3 +548,112 @@ TEST_F(TestFilesystemFixture, calculate_file_size) {
     g_allocator.deallocate(non_existing_path, g_allocator.state);
   });
 }
+
+TEST_F(TestFilesystemFixture, directory_iterator) {
+  char * path =
+    rcutils_join_path(this->test_path, "dummy_folder", g_allocator);
+  OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT(
+  {
+    g_allocator.deallocate(path, g_allocator.state);
+  });
+
+  rcutils_dir_iter_t * iter = rcutils_dir_iter_start(path, g_allocator);
+  ASSERT_NE(nullptr, iter);
+  OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT(
+  {
+    rcutils_dir_iter_end(iter);
+  });
+
+  const char * expected[] = {
+    ".",
+    "..",
+    "dummy.dummy",
+  };
+
+  do {
+    size_t i;
+    for (i = 0; i < sizeof(expected) / sizeof(expected[0]); i++) {
+      if (nullptr != expected[i] && 0 == strcmp(expected[i], iter->entry_name)) {
+        expected[i] = nullptr;
+        break;
+      }
+    }
+
+    if (i >= sizeof(expected) / sizeof(expected[0])) {
+      ADD_FAILURE() << "Unexpected entry " << iter->entry_name << " during iteration";
+    }
+  } while (rcutils_dir_iter_next(iter));
+
+  for (size_t i = 0; i < sizeof(expected) / sizeof(expected[0]); i++) {
+    EXPECT_EQ(nullptr, expected[i]) << "Expected entry " << expected[i] << " not found";
+  }
```

Comment:
This is a pretty terrible way to do this, but the enumeration order is non-deterministic. The case should be well-covered by `list_directory` if we decide to keep that function, so I might just drop this test.

Reply:
Fixed in 1242828